core/migration/coordinator.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package migration
+
+import (
+	"context"
+
+	"github.com/juju/description/v4"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+)
+
+// BaseOperation is a base implementation of the Operation interface.
+// The rollback operation is a no-op by default.
+type BaseOperation struct{}
+
+// Rollback is a no-op by default.
+func (b *BaseOperation) Rollback(context.Context) error {
+	return nil
+}
+
+// Operation is a single step in a migration.
+type Operation interface {
+	// Setup is called before the operation is executed. It should return an
+	// error if the operation cannot be performed.
+	Setup(Scope) error
+
+	// Execute is called to perform the operation. It should return an error
+	// if the operation fails.
+	Execute(context.Context, description.Model) error
+
+	// Rollback is called if the operation fails. It should attempt to undo
+	// any changes made by the operation. This is best effort, and may not
+	// always be possible.
+	Rollback(context.Context) error
```

Comment:
We should have a statement here that rollback logic should only ever be needed on operations that use the controller scope, because for an aborted model migration we should just be nuking the DB.

Perhaps also note that the only reason we need it is that modifications to multiple DBs makes transactionality at the driver level impossible.